### PR TITLE
[CALCITE-3437] ElasticSearch adapter. Support MatchQuery in elasticse…

### PR DESCRIPTION
**Pull request by shlok7296**
…arch

### CODE REVIEW

**Comment by asereda-gs** [comment]
@shlok7296 thank you for this contribution. Please add your full name in commit message:
> [CALCITE-1234] description (FirstName LastName)

Is it possible to add (integration) tests querying directly elastic ? See `elasticsearch/test` folder for examples.

**Comment by shlok7296** [comment]
@asereda-gs - Plz provide a sample query or reference test-case for sql 'Contains' query. 
I prepared filters using RexBuilder for which it is working.

**Comment by asereda-gs** [comment]
@shlok7296 pls take a look at [ElasticSearchAdapterTest](https://github.com/apache/calcite/blob/master/elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java). 

```java
public void example() {
calciteAssert()
        .query("select city from zips where city contains 'YORK' ")
        .returnsUnordered("state=NY; city=NEW YORK",
// ... 
          );
}
```

You can validate both query being sent to elastic (via `.queryContains()`) as well as JSON result (using `returnsUnordered()`).

**Comment by shikhasomani** [comment]
@asereda-gs Calcite Contains works for datetime range (periods) as tested in [SqlValidatorTest](https://github.com/apache/calcite/blob/master/core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java#L7290). 
Below query when tested on postgres via Calcite (jdbc adapter), it gives error:

Query: select city from zips where city contains 'YORK'
Error: Cannot apply 'CONTAINS' to arguments of type '<VARCHAR(200)> CONTAINS <CHAR(15)>'. Supported form(s): '<DATETIME> CONTAINS <DATETIME>'

**Comment by asereda-gs** [comment]
Can you show how you're executing it with RexBuilder ?

**Comment by shikhasomani** [comment]
@asereda-gs We have added unit test for ElasticSearch Match query. It prepares CONTAINS query using RelBuilder, RexBuilder which is translated to ES match query. It finally runs query on text field (city) which is having tokenized data and returns matching data.

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java [hunk 1]
```diff
@@ -705,6 +737,97 @@ public void approximateCount() throws Exception {
...
+        final String result = ""
+            + "_MAP={id=02154, city=NORTH WALTHAM, loc=[-71.236497, 42.382492], pop=57871, state=MA}; city=NORTH WALTHAM\n";
+
+        //Validate query prepared
+        assertThat(root, hasTree(builderExpected));
+
+        //Validate result returned from ES
+        assertThat(s, is(result));
+      }
+    } catch (Exception e) {
```
Comment (asereda-gs) [inline elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: You don't need to wrap the exception. Let it propagate and be reported by JUnit

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java [hunk 2]
```diff
@@ -705,6 +737,97 @@ public void approximateCount() throws Exception {
...
+
+      String builderExpected = ""
+          + "LogicalFilter(condition=[CONTAINS($1, 'waltham')])\n"
+          + "  LogicalProject(_MAP=[$0], city=[ITEM($0, 'city')])\n"
+          + "    LogicalTableScan(table=[[elastic, zipswithtext]])\n";
+
+      RelNode root = builder.build();
+
+      RelRunner ru = (RelRunner) con.unwrap(Class.forName("org.apache.calcite.tools.RelRunner"));
+      try (PreparedStatement preparedStatement = ru.prepare(root)) {
```
Comment (asereda-gs) [inline elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: Use directly `RelRunners.run(RelNode)`

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java [hunk 3]
```diff
@@ -389,7 +420,8 @@ public void limit2() {
                 "size:5"));
   }
 
-  @Test public void testFilterSort() {
+  //@Test
```
Comment (asereda-gs) [inline elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: Why remote this (`testFilterSort`) test ?

File: elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/PredicateAnalyzer.java [hunk 4]
```diff
@@ -543,6 +548,8 @@ public boolean isPartial() {
     /**
```
Comment (asereda-gs) [inline elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/PredicateAnalyzer.java]: This javadoc belongs to `not()` method

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java [hunk 5]
```diff
@@ -705,6 +737,97 @@ public void approximateCount() throws Exception {
             "state=AL; EXPR$1=3; EXPR$2=3",
             "state=AR; EXPR$1=3; EXPR$2=3");
   }
+  /**
+   * Test the ElasticSearch match query. The match query is translated from CONTAINS query which
```
Comment (asereda-gs) [inline elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: I didn't expect this test to be so long. Can you pls move it into separate class (similarly to `AggregationTest`) ? There you can create any mapping you want

File: core/src/test/java/org/apache/calcite/test/CalciteAssert.java [hunk 6]
```diff
@@ -672,7 +672,7 @@ static void assertPrepare(
   }
 
   /** Converts a {@link ResultSet} to a string. */
-  static String toString(ResultSet resultSet) throws SQLException {
+  public static String toString(ResultSet resultSet) throws SQLException {
```
Comment (asereda-gs) [inline core/src/test/java/org/apache/calcite/test/CalciteAssert.java]: Not sure this method was intended to be public, but right now, can't think of a better alternative.

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java (hunk 2 above)
Comment (shikhasomani) [reply elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: Using RelRunners.run(RelNode) doesn't work, hence using RelRunner only.

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java (hunk 1 above)
Comment (shlok7296) [reply elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: Done

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java (hunk 3 above)
Comment (shlok7296) [reply elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: Done, reverted this change.

File: elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/PredicateAnalyzer.java (hunk 4 above)
Comment (shlok7296) [reply elasticsearch/src/main/java/org/apache/calcite/adapter/elasticsearch/PredicateAnalyzer.java]: Done

File: elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java (hunk 5 above)
Comment (shlok7296) [reply elasticsearch/src/test/java/org/apache/calcite/adapter/elasticsearch/ElasticSearchAdapterTest.java]: Done. Moved it to new class MatchTest

**Comment by asereda-gs** [comment]
Thank you @shlok7296  for this PR

**Comment by houbb** [comment]
@asereda-gs @shlok7296 

In calcite-elasticsearch v1.36.0，I wanna to CONTAINS to search ES，but got error。

It seems CONTAINS not support varchar type by default. How to fix it? thx


```java
private static void showContains(CalciteConnection calciteConnection) {
    try {
        String sql = "SELECT * FROM booksmapping WHERE title contains 'Java'";
        Statement statement = calciteConnection.createStatement();
        ResultSet resultSet = statement.executeQuery(sql);
        System.out.println(ResultSetUtil.resultString(resultSet));
    } catch (SQLException e) {
        throw new RuntimeException(e);
    }
}
```

error logs

```
Caused by: org.apache.calcite.sql.validate.SqlValidatorException: Cannot apply 'CONTAINS' to arguments of type '<VARCHAR> CONTAINS <CHAR(4)>'. Supported form(s): '<DATETIME> CONTAINS <DATETIME>'
'<DATETIME> CONTAINS <DATETIME>'
'<DATETIME> CONTAINS <INTERVAL>'
'<DATETIME> CONTAINS <INTERVAL>'
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.apache.calcite.runtime.Resources$ExInstWithCause.ex(Resources.java:507)
	at org.apache.calcite.runtime.Resources$ExInst.ex(Resources.java:601)
	... 37 more
```

**Pull request merged**